### PR TITLE
Feature/#1763 local replays pagination

### LINKED ISSUE #1078: Don't cap local replay list after 300 files
Right now there is a hard limit of 300 local replays to show.

https://github.com/FAForever/downlords-faf-client/blob/d5cf40884953a578858c01c3ff0ca9843a82189e/src/main/java/com/faforever/client/replay/ReplayService.java#L170

There needs to be a better we of preventing OutOfMemoryExceptions with some sort of paginated loading.

### LINKED ISSUE #1763: Pagination of local replays
Some users have more than 10.000 local replays, that will be loaded on invocation 
This can cause an error.
Load only the 100 newest and add a pagination with 100 items per page.

### PR DESCRIPTION
**Pull request by jofaul**
Done after unittesting

### CODE REVIEW

**Comment by micheljung** [comment]
It looks like there are many commits in this PR that shouldn't be. Please sort this out first.

**Comment by 1-alex98** [comment]
@micheljung on the other hand I can also squash this on merge

File: src/main/java/com/faforever/client/replay/ReplayService.java [hunk 1]
```diff
@@ -139,19 +142,27 @@
   private final ExecutorService executorService;
   private Thread directoryWatcherThread;
   private WatchService watchService;
-  protected List<Replay> localReplays = new ArrayList<Replay>();
+  protected List<Replay> localReplays = new ArrayList<>();
 
-  public void startLoadingAndWatchingLocalReplays() {
+  public void startLoadingAndWatchingLocalReplays(Pagination pagination) {
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/replay/ReplayService.java]: Passing an UI Elements to a service should never happen!

File: src/main/java/com/faforever/client/replay/ReplayService.java [hunk 2]
```diff
@@ -139,19 +142,27 @@
...
+      logger.error("Failed loading total amount of replays", e);
+    }
+
     LoadLocalReplaysTask loadLocalReplaysTask = applicationContext.getBean(LoadLocalReplaysTask.class);
-    taskService.submitTask(loadLocalReplaysTask).getFuture().thenAccept( replays -> {
+    taskService.submitTask(loadLocalReplaysTask).getFuture().thenAccept(replays -> {
       localReplays.clear();
       localReplays.addAll(replays);
-      publisher.publishEvent(new LocalReplaysChangedEvent(this, replays, new ArrayList<Replay>()));
+      publisher.publishEvent(new LocalReplaysChangedEvent(this, replays, new ArrayList<>()));
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/replay/ReplayService.java]: Publish page count here maybe...

File: src/main/java/com/faforever/client/replay/ReplayService.java [hunk 3]
```diff
@@ -139,19 +142,27 @@
...
     if (Files.notExists(replaysDirectory)) {
       noCatch(() -> createDirectories(replaysDirectory));
     }
 
+    String replayFileGlob = clientProperties.getReplay().getReplayFileGlob();
+    try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(replaysDirectory, replayFileGlob)) {
+      final int pageCount = (int) Math.ceil((StreamSupport.stream(directoryStream.spliterator(), false).count() / (double) REPLAYS_PER_PAGE));
+      Platform.runLater(() -> pagination.setPageCount(pageCount));
+    } catch (IOException e) {
+      logger.error("Failed loading total amount of replays", e);
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/replay/ReplayService.java]: This is not an error, rather warning I say

File: src/main/java/com/faforever/client/replay/ReplayService.java [hunk 4]
```diff
@@ -186,8 +207,8 @@ protected Thread startDirectoryWatcher(Path replaysDirectory) throws IOException
 
   @VisibleForTesting
   protected void onLocalReplaysWatchEvent(WatchKey key) {
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/replay/ReplayService.java]: Have u thought about what happens if a user drops a new replay into his replay folder... that is what happens here.

Comment (1-alex98) [reply src/main/java/com/faforever/client/replay/ReplayService.java]: Maybe u can simplify this method by just going back to the first page :thinking:

**Review by 1-alex98 — CHANGES_REQUESTED** [review]
Also misses unit tests

File: src/main/java/com/faforever/client/replay/ReplayService.java (hunk 2 above)
Comment (1-alex98) [reply src/main/java/com/faforever/client/replay/ReplayService.java]: I am surprised this works with an event publisher. Probably the Controller a singleton and that's why it works. Whatever let's leave it like this.

**Comment by micheljung** [comment]
> @micheljung on the other hand I can also squash this on merge

If you write a proper commit message, sure. But I won't accept the default generated commit message ;-)

File: src/main/java/com/faforever/client/replay/ReplayService.java [hunk 5]
```diff
@@ -139,19 +140,29 @@
   private final ExecutorService executorService;
   private Thread directoryWatcherThread;
   private WatchService watchService;
-  protected List<Replay> localReplays = new ArrayList<Replay>();
+  protected List<Replay> localReplays = new ArrayList<>();
+  private double localReplaysItemsSize;
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/replay/ReplayService.java]: whys that a double

File: src/main/java/com/faforever/client/replay/ReplayService.java [hunk 6]
```diff
@@ -134,36 +134,47 @@
...
   private final ModService modService;
   private final MapService mapService;
-  private final ApplicationEventPublisher publisher;
+  private final EventBus eventBus;
   private final MapGeneratorService mapGeneratorService;
-  private final ExecutorService executorService;
   private Thread directoryWatcherThread;
-  private WatchService watchService;
-  protected List<Replay> localReplays = new ArrayList<Replay>();
+  protected List<Replay> localReplays = new ArrayList<>();
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/replay/ReplayService.java:129]: Written to but never read, right? If so remove

**Review by 1-alex98 — CHANGES_REQUESTED** [review]

File: src/main/java/com/faforever/client/replay/ReplayService.java [hunk 7]
```diff
@@ -134,36 +134,47 @@
...
+    reloadLocalReplays();
   }
 
-  public Collection<Replay> getLocalReplays() {
-    return localReplays;
+  private void reloadLocalReplays() {
+    Path replaysDirectory = preferencesService.getReplaysDirectory();
+    pageCountLocalReplays = 1;
+    String replayFileGlob = clientProperties.getReplay().getReplayFileGlob();
+    try (DirectoryStream<Path> directoryStream = Files.newDirectoryStream(replaysDirectory, replayFileGlob)) {
```
Comment (1-alex98) [inline src/main/java/com/faforever/client/replay/ReplayService.java]: Files.list

**Review by 1-alex98 — CHANGES_REQUESTED** [review]

**Comment by Sheikah45** [comment]
Did we want to do the local replay pagination this way? I can rebase and bring it up to current.

**Comment by 1-alex98** [comment]
We need to definetely rebase it. How would u do it otherwise?

**Comment by Sheikah45** [comment]
I would probably want to refactor it a bit to make it more similar to the other paginations as much as possible

**Comment by 1-alex98** [comment]
ok but we can definitely not use the same code base

**Comment by Sheikah45** [comment]
Maybe not. Also do you think that we can't just make a card for local replays?

**Comment by Sheikah45** [comment]
I also meant more so in the style and naming conventions

**Comment by 1-alex98** [comment]
Have fun do what u like :D

**Comment by Sheikah45** [comment]
Fixes #1763

**Review by 1-alex98 — COMMENTED** [review]
Looks neat

**Pull request merged**